File: AntBot/src/antbot/antbot_bringup/script/real/odom/odom_publisher.py
#!/usr/bin/env python
#coding=utf-8

import roslib; roslib.load_manifest('antbot_bringup')
#添加此句给pythonimport 时 寻找其他的文件用来
import sys

spath=sys.path[0]
snum=spath.rfind('/')
ssyspath=spath[0:snum]
sys.path.append(ssyspath)

import rospy
import tf
from agv.msg            import Num, carOdom, MImu
from geometry_msgs.msg  import Twist, Quaternion, TransformStamped
from math               import sin, cos ,pi
from sensor_msgs.msg    import JointState
from nav_msgs.msg       import Odometry
from tf.transformations import euler_from_quaternion
from tf.transformations import quaternion_from_euler

class PublishOdom():
    def __init__(self):
        #初始化节点
        rospy.init_node("PublishOdom", anonymous = True)
        
        #结束节点时执行nodeshutdown函数
        rospy.on_shutdown(self.nodeShutdown)
        
        #订阅car_speed话题
        rospy.Subscriber("/car_speed", carOdom, self.speed_callback)
        
        #订阅速度话题
        rospy.Subscriber("/cmd_vel", Twist, self.cmd_callback)

        #订阅imu话题
        rospy.Subscriber("/rawImu_message", MImu, self.imu_callback)
        
        #关节坐标发布者
        self.joint_pub = rospy.Publisher("/joint_state", JointState, queue_size = 20)
        
        #里程信息发布者
        self.odom_pub = rospy.Publisher("/odom", Odometry, queue_size = 20)
        
        self.odom_broadcaster = tf.TransformBroadcaster()
        self.imu_msg          = MImu()
        self.cmd_msg          = Twist()
        self.xPosition  = 0.0
        self.yPosition  = 0.0
        self.thPosition = 0.0
        self.current_odom_msg    = carOdom()
        self.last_odom_msg       = carOdom()
        self.last_angle_z    = 0.0
        self.current_angle_z = 0.0
        
        #编码轮的系数
        k_x = 0.872
        k_y = 0.874

        #编码轮半径
        radius = 0.058

        #计算位置的系数
        self.pk_x = 2 * pi * radius * k_x / 2400.0
        self.pk_y = 2 * pi * radius * k_y / 2400.0
            
    #速度回调函数
    def speed_callback(self, msg):

        #得到当前的航向角
        self.current_angle_z = self.imu_msg.angle_z
        
        #读取x，y轴的脉冲数
        self.current_odom_msg = msg
        
        #单位时间的脉冲数
        diff_x = self.current_odom_msg.xOdom - self.last_odom_msg.xOdom
        diff_y = self.current_odom_msg.yOdom - self.last_odom_msg.yOdom
        
        #特殊情况
        if diff_x > 3000:
            diff_x = diff_x - 4800
        elif diff_x < -3000:
            diff_x = diff_x + 4800
    
        if diff_y > 3000:
            diff_y = diff_y - 4800
        elif diff_y < -3000:
            diff_y = diff_y + 4800
        
        #机器人坐标上单位时间内的位姿变化
        i = diff_x * self.pk_x
        j = diff_y * self.pk_y

        #世界坐标上d单位时间内的位姿变化
        delta_th   = self.current_angle_z - self.last_angle_z
        
        """
            判断是否有旋转运动如果有，这里只有两种运动形式，平移和旋转。
        """
        #平移
        if not self.cmd_msg.angular.z == 0:
            delta_x = 0.0
            delta_y = 0.0    
        #旋转
        else:
            delta_x = i * cos(self.thPosition) - j * sin(self.thPosition)
            delta_y = i * sin(self.thPosition) + j * cos(self.thPosition)
        
        #累计的位置
        self.xPosition  += delta_x
        self.yPosition  += delta_y
        self.thPosition += delta_th
    
        #发布关节的坐标关系
        joint_state = JointState()
        joint_state.header.stamp = rospy.Time.now()
        joint_state.name.append("base_to_wheel1")
        joint_state.position.append(0)
        joint_state.name.append("base_to_wheel2")
        joint_state.position.append(1)
        joint_state.name.append("base_to_wheel3")
        joint_state.position.append(2)
        joint_state.name.append("base_to_wheel4")
        joint_state.position.append(3)
        self.joint_pub.publish(joint_state)
        
        #发布TF关系
        odom_quat = Quaternion()
        odom_quat = quaternion_from_euler(0, 0, self.thPosition)
        self.odom_broadcaster.sendTransform((self.xPosition, self.yPosition, 0),
                                            odom_quat,
                                            rospy.Time.now(),
                                            "odom",
                                            "base_link")
        #发布航迹推演消息
        #定义额一个Odometry消息
        odom = Odometry()
        odom.header.stamp = rospy.Time.now()
        odom.header.frame_id = "odom"

        #设置位姿
        odom.pose.pose.position.x = self.xPosition
        odom.pose.pose.position.y = self.yPosition
        odom.pose.pose.position.z = 0.0
        odom.pose.pose.orientation.x = odom_quat[0]
        odom.pose.pose.orientation.y = odom_quat[1]
        odom.pose.pose.orientation.z = odom_quat[2]
        odom.pose.pose.orientation.w = odom_quat[3]

        #设置速度
        odom.child_frame_id = "base_link"
        odom.twist.twist.linear.x = 0#self.vx
        odom.twist.twist.linear.y = 0#self.vy
        odom.twist.twist.angular.z = 0#0self.vth

        #协方差
        odom.pose.covariance[0]  = 0.1
        odom.pose.covariance[7]  = 0.1
        odom.pose.covariance[35] = 0.2
        odom.pose.covariance[14] = 10000000 #set a non-zero covariance on unused
        odom.pose.covariance[21] = 10000000 #dimensions (z, pitch and roll); this
        odom.pose.covariance[28] = 10000000 #is a requirement of robot_pose_ekf
        self.odom_pub.publish(odom)
        
        #迭代
        self.last_odom_msg = self.current_odom_msg
        self.last_angle_z  = self.current_angle_z
        
    #速度回调函数
    def cmd_callback(self, msg):
        #读取旋转的角速度
        self.cmd_msg.angular.z = msg.angular.z
   
    #imu回调函数
    def imu_callback(self, msg):
        #读取航向角
        self.imu_msg.angle_z = msg.angle_z
        print "angle.z: %f"%(msg.angle_z / (pi) * 180)
                   
    #节点结束时的执行函数
    def nodeShutdown(self):
        print "PublishOdom node is shutdown"

if __name__=="__main__":
    try:
        print "being PublishOdom..."
        PublishOdom()
        rospy.spin()
    except:
        rospy.loginfo("PublishOdom node terminated !")
        

File: AntBot/src/antbot/antbot_bringup/src/fake/odom_publisher.cpp
/*
订阅机器人的速度，发布轮子的Joint信息
发布TF关系
*/
#include <ros/ros.h>
#include <tf/transform_broadcaster.h>
#include <nav_msgs/Odometry.h>
#include <antbot_bringup/antbotOdom.h>
#include <sensor_msgs/JointState.h>

class PublishOdom
{
private:
    ros::NodeHandle n;
    ros::Publisher odom_pub;
    ros::Subscriber antbotspeed_sub;
    ros::Time current_time,last_time;
    ros::Publisher joint_pub;
    tf::TransformBroadcaster odom_broadcaster;
    double x_sum;
    double y_sum;
    double th_sum;
    float L;//antbot半
    float W;//antbot半宽
    float R;//antbot轮子的半径
    float v[4];
    float p[4];
    
public:
    PublishOdom(){
        odom_pub = n.advertise<nav_msgs::Odometry>("odom", 1);
        joint_pub = n.advertise<sensor_msgs::JointState>("wheel_joint_states", 1);
        antbotspeed_sub = n.subscribe("antbot_speed", 1, &PublishOdom::sub_callrear, this);
        L = 0.4;
        W = 0.3;
        R = 0.12;
    }
    void sub_callrear(const antbot_bringup::antbotOdom::ConstPtr& input)
     {
         current_time=ros::Time::now();
         float vx = input->vx;//the x speed of the antbot
         float vy = input->vy;//the y speed of the antbot
         float vth = input->vth;//the thea speed of the antbot
         //compute odometry in a typical way given the velocities of the robot
         double dt = (current_time - last_time).toSec();
         double delta_x  = vx * dt; //(vx * cos(th_sum) - vy * sin(th_sum)) * dt;
         double delta_y  = vy * dt; //(vx * sin(th_sum) + vy * cos(th_sum)) * dt;
         double delta_th = vth * dt;
         th_sum += delta_th;
         printf("th_sum : %f\n", th_sum);
         x_sum  += delta_x * cos(th_sum) - delta_y * sin(th_sum);
         y_sum  += delta_x * sin(th_sum) + delta_y * cos(th_sum);
         
         
         v[0] = vx + vy + (L + W) * vth;
         v[1] = vx - vy - (L + W) * vth;
         v[2] = vx + vy - (L + W) * vth;
         v[3] = vx - vy + (L + W) * vth;


		 for(int i = 0; i < 4; i++)
	         p[i] += (v[i]/R)*dt;
	     

         sensor_msgs::JointState joint_state;
         //update joint_state
         joint_state.header.stamp = ros::Time::now();
         joint_state.name.resize(4);
         joint_state.position.resize(4);
         joint_state.name[0] ="front_right_wheel_to_base_link";
         joint_state.position[0] = p[0];
         joint_state.name[1] ="front_left_wheel_to_base_link";
         joint_state.position[1] = p[1];
         joint_state.name[2] ="rear_left_wheel_to_base_link";
         joint_state.position[2] = p[2];
         joint_state.name[3] ="rear_right_wheel_to_base_link";
         joint_state.position[3] = p[3];
         //send the joint state and transform
         joint_pub.publish(joint_state);


         //since all odometry is 6DOF we'll need a quaternion created from yaw
         /* 将里程计的偏航角转换成四元数，四元数效率高，这样使用二维和三维的功能包是一样的.*/
        geometry_msgs::Quaternion odom_quat = tf::createQuaternionMsgFromYaw(th_sum);
        //first, we'll publish the transform over tf
        geometry_msgs::TransformStamped odom_trans;
        odom_trans.header.stamp = current_time;
        odom_trans.header.frame_id = "odom";
        odom_trans.child_frame_id = "base_link";
        odom_trans.transform.translation.x = x_sum;
        odom_trans.transform.translation.y = y_sum;
        odom_trans.transform.translation.z = 0.0;
        odom_trans.transform.rotation = odom_quat;
        //send the transform
        odom_broadcaster.sendTransform(odom_trans);


         //next, we'll publish the odometry message over ROS
        nav_msgs::Odometry odom;
        odom.header.stamp = current_time;
        odom.header.frame_id = "odom";
        //set the position
        odom.pose.pose.position.x = x_sum;
        odom.pose.pose.position.y = y_sum;
        odom.pose.pose.position.z = 0.0;
        odom.pose.pose.orientation = odom_quat;
        //set the velocity
        odom.child_frame_id = "base_link";
        odom.twist.twist.linear.x = vx;
        odom.twist.twist.linear.y = vy;
        odom.twist.twist.angular.z = vth;

        //publish the message
        odom.pose.covariance[0]  = 0.1;
        odom.pose.covariance[7]  = 0.1;
        odom.pose.covariance[35] = 0.2;

        odom.pose.covariance[14] = DBL_MAX; // set a non-zero covariance on unused
        odom.pose.covariance[21] = DBL_MAX; // dimensions (z, pitch and roll); this
        odom.pose.covariance[28] = DBL_MAX; // is a requirement of robot_pose_ekf
        odom_pub.publish(odom);

        last_time = current_time;
     }

};
int main(int argc,char** argv)
{
    ros::init(argc,argv,"odometry_publiser");
    // topic you want to publish
    PublishOdom PublishOdomObject;
    
    ros::spin();
    return 0;
}
